### PR TITLE
Import logs migrations from travis-logs app

### PR DESCRIPTION
**Pull request by igorwwwwwwwwwwwwwwwwwwww**
The status quo is that migrations are split between `travis-migrations` (this repo) for the main migrations and [travis-logs](https://github.com/travis-ci/travis-logs) for the logs migrations.

Some apps, e.g. [travis-api](https://github.com/travis-ci/travis-api) require the logs database to be present in order to operate in development and to run the test suite. The `travis-migrations` repo contains a database dump that can be imported directly for those use cases, without having to load a bunch of migrations dependencies.

The logs database used to have its schema maintained in this repo as well as the travis-logs repo. And the logs tables were available as part of the `structure.sql` dump. That was changed recently ([commit](https://github.com/travis-ci/travis-migrations/commit/ebe0d03543841a6b2edbf014ca7ce19374ba7a38)).

A consequence of that change is that travis-api and other apps can no longer safely use the master branch of travis-migrations. Instead, they need to depend on an old version in which the logs tables are still present. This will lead to a big merge mess when the next schema change happens, and effectively forces us to maintain a separate branch of travis-migrations with the logs tables added that is only used during development and for running tests.

The initial proposal was to roll back the commit that removed logs migrations, but this is not really an option, because then travis-migrations would not mirror our production setup. In production we have two separate databases for main and logs. In fact, completely separate postgres instances.

Long story short:

So instead of trying to make travis-migrations support a single database, this PR attempts to make travis-migrations model our production setup, and also make it easier for apps that require that database to run a similar setup in their tests. Thus everything is closer to how we have it in production, and also more likely to catch weird bugs, like joining across database instances.

See [the sibling PR on travis-api](https://github.com/travis-ci/travis-api/pull/350) for an example of how this change affects other apps.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -15,6 +17,48 @@ task :default => [:spec]
 
 Rake::Task["db:structure:dump"].clear unless Rails.env.development?
 
+def connect_sequel
```
Comment (rkh) [inline Rakefile]: should probably be `def self.connect_sequel`, otherwise this method is defined on `Object`.

File: Rakefile [hunk 2]
```diff
@@ -15,6 +17,48 @@ task :default => [:spec]
...
+  if ENV['LOGS_DATABASE_URL'].nil?
+    raise 'No logs database URL set, you must set LOGS_DATABASE_URL env var'
```
Comment (renee-travisci) [inline Rakefile]: The logs database isn't required everywhere - can we make this a warning instead?

**Comment by svenfuchs** [comment]
Looks good to me. However, I think it also could use some more work.

For tests to be run on Travis CI most apps (such as Gatekeeper, Hub, ..., as well as API at some point) we pull the schema via HTTP. So the schema (`structure.sql`) needs to include the logs tables.

I suppose dumping the schema might be possible, but also a little cumbersome (needs to set the logs database url and such).

I am also wondering if it would not be worth converting this from Sequel to AR at this point already.

Also, personally, I'd put the migrations into two sub directories named by the database. E.g. `./db/main` and `./db/logs`.

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
@svenfuchs @renee-travisci  is there a reason for pulling in the schema via HTTP? would it be possible to make travis-migrations a dependency and pull it from disk, and then run the migrations similarly to the way they are in API?

**Comment by svenfuchs** [comment]
@igorwwwwwwwwwwwwwwwwwwww yes, making it a dependency adds several gems that aren't needed for anything but running the tests on Travis CI. E.g. Sync uses Sequel, and doesn't need ActiveRecord. We've just moved away from bundling travis-migrations, and I'd prefer being able to keep that usecase.

File: Gemfile [hunk 3]
```diff
@@ -8,3 +8,5 @@ gem 'data_migrations', '~> 0.0'
 gem 'bundler', '~> 1.7'
 gem 'rspec', '~>3.4'
 gem 'rails_12factor'
+
+gem 'sequel'
```
Comment (renee-travisci) [inline Gemfile]: I think we need to add this to the gemspec so bundler picks it up as a runtime_dependancy for using this gem

Comment (igorwwwwwwwwwwwwwwwwwwww) [inline Gemfile]: Gotcha! :)

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Major update:

* logs migrations have been rewritten in activerecord
* there are two separate directories for main vs logs
* switching between the two is controlled by a `LOGS_DATABASE` environment variable
* the logs database schema is dumped to an sql file, so it can be loaded directly by other apps

File: Rakefile [hunk 4]
```diff
@@ -5,6 +5,14 @@ require "micro_migrations"
 
 ActiveRecord::Base.schema_format = :sql
 
+if ENV['LOGS_DATABASE'] == '1'
```
Comment (renee-travisci) [inline Rakefile]: Let's make these 2 separate rake tasks. One can be rake db:migrate the other can be rake db:migrate:logs. Then we can have 2 tests one for each rake task, and save on the env var. Also, I would add a third task that can be used rake db:migrate:all which calls both other rake tasks.

File: .travis.yml [hunk 5]
```diff
@@ -27,4 +28,4 @@ install:
   - bundle _1.12.4_ install --jobs=3 --retry=3 --path=${BUNDLE_PATH:-vendor/bundle}
 
 before_script:
-  - 'cp config/database.yml.travis config/database.yml'
+  - psql -c 'create database travis_ci_test;' -U postgres
```
Comment (renee-travisci) [inline .travis.yml]: We should keep the config so the tests are as close to what people run in dev/prod as possible and if a name changes this doesn't get lost. Also the tests use the database.yml to make the connection.

File: spec/travis_migrations_spec.rb [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
-
 describe 'Rake tasks' do
   let(:config) { YAML.load(ERB.new(File.read('config/database.yml')).result) }
   let(:conn)   { ActiveRecord::Base.connection }
   let(:tables) { conn.select_values("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'") }
   before       { ActiveRecord::Base.establish_connection(config['test']) }
 
   def run(cmd)
-    system "RAILS_ENV=test bundle exec #{cmd}"
+    system "RAILS_ENV=test LOGS_DATABASE=#{ENV['LOGS_DATABASE'].to_i} bundle exec #{cmd}"
```
Comment (renee-travisci) [inline spec/travis_migrations_spec.rb]: This only tests the logs tables from what I can see

Comment (igorwwwwwwwwwwwwwwwwwwww) [inline spec/travis_migrations_spec.rb]: It inherits the `LOGS_DATABASE` env var from `.travis.yml` and passes it through to the sub-process -- both are covered.

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Making two separate commands is closer to what I was trying to do in the beginning. It is not super trivial though -- it would basically require having two separate config files `database.main.yml` and `database.logs.yml`. Is that acceptable?

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Another thing. We are currently importing the existing rake tasks from active_record. I feel like it would take a significant effort to hack those in a way that supports more than one instance in the same process. In particular because it relies heavily on global state like `Rails.application.config`.

We would need to have `logs` versions of _all_ of these commands, which is not so nice imo:

```
➜  travis-migrations git:(igor-logs-migrations) ✗ rake -T
rake db:create                          # Creates the database from DATABASE_URL o...
rake db:drop                            # Drops the database from DATABASE_URL or ...
rake db:fixtures:load                   # Load fixtures into the current environme...
rake db:migrate                         # Migrate the database (options: VERSION=x...
rake db:migrate:status                  # Display status of migrations
rake db:rollback                        # Rolls the schema back to the previous ve...
rake db:schema:cache:clear              # Clear a db/schema_cache.dump file
rake db:schema:cache:dump               # Create a db/schema_cache.dump file
rake db:schema:dump                     # Create a db/schema.rb file that is porta...
rake db:schema:load                     # Load a schema.rb file into the database
rake db:seed                            # Load the seed data from db/seeds.rb
rake db:setup                           # Create the database, load the schema, an...
rake db:structure:dump                  # Dump the database structure to db/struct...
rake db:structure:load                  # Recreate the databases from the structur...
rake db:version                         # Retrieves the current schema version number
```

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
@renee-travisci merged your PR, awaiting final review!

**Comment by renee-travisci** [comment]
Should we mention in the README that if you are using 1 database locally, and you want to update or change log tables, you will need to update the database.yml (or your ENV vars) here to have both logs and main point to the same DB? (Enterprise will need to do that, if they don't already use ENV vars)

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
I would be curious to know if enterprise uses a single database, or two separate databases within the same postgres instance. (perhaps @bnferguson can confirm?)

If the latter is the case, I would try and make that the setup we use everywhere, and update the README accordingly.

**Comment by bnferguson** [comment]
> I would be curious to know if enterprise uses a single database, or two separate databases within the same postgres instance. (perhaps @bnferguson can confirm?)

Currently it uses a single db. But I don't know if I'd rely on that always being the case because in some larger installations we have seen some performance issues (I'm guessing similar to the ones we've seen in .com/.org that led to split dbs?).

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Another thing we need to do upon merge is migrate staging and prod to the new migrations setup. Essentially a schema migration for our schema migrations:

```
DROP TABLE schema_migrations;
CREATE TABLE schema_migrations (
    version character varying NOT NULL
);
CREATE UNIQUE INDEX unique_schema_migrations ON schema_migrations USING btree (version);

INSERT INTO schema_migrations (version) VALUES ('20161116162355');
INSERT INTO schema_migrations (version) VALUES ('20161116162356');
```

**Comment by bnferguson** [comment]
Want to give this a review for Enterprise but am missing the overall context of this move and why it's happening since the PR description is empty. Mind giving an overview as to what this change is and why it needs to happen? 

(As a side note: this is something that also helps when writing release notes/change logs for feature releases in Enterprise - helps us understand what was added/fixed/changed without having to dive into comments and code. So anytime people can fill that out it helps immensely for those of us that are missing the context that might live other places. :heart: )

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
@bnferguson I've added an extensive description to the PR, I hope that helps!

**Comment by bnferguson** [comment]
@igorwwwwwwwwwwwwwwwwwwww OMG, this is _so_ good! Thank you so much. ❤️  (your description actually made some other conversations I've been around much clearer too)

Getting on with the review now!

File: README.md [hunk 7]
```diff
@@ -72,6 +66,19 @@ Append `HEAD:master` to the pit push if you are on a branch and want to push tha
 git push git@heroku.com:<app>.git HEAD:master
 ```
 
+Logs database
+-------------
+
+To run the migration commands for the logs database, set `RAILS_ENV=development_logs` as an environment variable.
+
```
Comment (bnferguson) [inline README.md]: So, if I'm understanding this correctly,  it looks like this is something I'll need to account for here: https://github.com/travis-pro/travis-enterprise/blob/master/src/main/bin/te-db-migrate (add a second line that changes the `RAILS_ENV` to `development_logs`)?

Will these migrations still come with `travis-api` (or in the case of enterprise `travis-pro-api`) or will I need to package this application separately to run the migrations?

File: config/database.yml [hunk 8]
```diff
@@ -8,11 +8,24 @@ defaults: &defaults
...
 
 development:
   <<: *defaults
   database: <%= ENV.fetch('DATABASE_NAME', 'travis_development') %>
+  eager_load: false
+
+development_logs:
+  <<: *defaults
+  database: <%= ENV.fetch('LOGS_DATABASE_NAME', 'travis_logs_development') %>
+  eager_load: false
```
Comment (bnferguson) [inline config/database.yml]: Looks like this should work for us in enterprise. 👍

File: README.md (hunk 7 above)
Comment (igorwwwwwwwwwwwwwwwwwwww) [inline README.md]: @bnferguson that is correct, you would need another line.

And I would recommend running the migrations separately form within the migrations project, as opposed to running them from API. Since that matches what we do in production.

Comment (bnferguson) [inline README.md]: > And I would recommend running the migrations separately form within the migrations project, as opposed to running them from API. Since that matches what we do in production.

👍 Awesome, I'll look at starting to bundle these in 2.1 and run them from there.

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
I've added a patch for `travis-logs` that removes the old sequel based migrations: https://github.com/travis-ci/travis-logs/pull/59

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Steps taken for each env:

```
cd travis-migrations

heroku git:remote travis-pro-production
git push heroku master
heroku pg -a travis-pro-logs-production | grep -e '===\|Add-on'

heroku addons:attach <addon> --as LOGS_DATABASE

heroku pg:psql -a travis-pro-logs-production # paste schema_migrations sql

heroku run RAILS_ENV=production_logs bundle exec rake db:migrate:status
```
